### PR TITLE
Initial commit.

### PR DESCRIPTION
**Pull request by dmitrizagidulin**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,44 +1,110 @@
...
+```js
+// a different key pair than the previous example
+await keyPair.export({publicKey: true, privateKey: true})
+// ->
+ {
+  id: 'did:example:1234#z6LSjeJZaUHMvEKW7tEJXV4PrSm61NzxxHhDXF6zHnVtDu9g',
+  type: 'X25519KeyAgreementKey2020',
+  controller: 'did:example:1234',
+  publicKeyBase58: '8y8Q4AUVpmbm2VrXzqYSXrYcAETrFgX4eGPJoKrMWXNv',
+  privateKeyBase58: '95tmYuhqSuJqY77FEg78Zy3LFQ1cENxGv2wMvayk7Lqf'
```
Comment (dlongley) [inline README.md]: Should be using multibase.

File: README.md [hunk 2]
```diff
@@ -1,44 +1,110 @@
...
+
+Deserializing:
+
+```js
+// Loading public key only
+const keyPair = await X25519KeyAgreementKey2020.from({
+  id: 'did:example:1234#z6LSjeJZaUHMvEKW7tEJXV4PrSm61NzxxHhDXF6zHnVtDu9g',
+  type: 'X25519KeyAgreementKey2020',
+  controller: 'did:example:1234',
+  publicKeyBase58: '8y8Q4AUVpmbm2VrXzqYSXrYcAETrFgX4eGPJoKrMWXNv'
```
Comment (dlongley) [inline README.md]: multibase

**Review by dlongley — APPROVED** [review]
I did not look too closely at this -- only noting that the main change is to use multibase for key representations.

File: README.md (hunk 1 above)
Comment (dmitrizagidulin) [reply README.md]: Fixed, thanks.

File: lib/X25519KeyAgreementKey2020.js [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+}
+
+export class X25519KeyAgreementKey2020 extends LDKeyPair {
+  /**
+   * @param {object} options - Options hashmap.
+   * @param {string} options.controller - Controller DID or document url.
+   * @param {string} [options.id] - Key id, typically composed of controller
+   *   URL and key fingerprint as hash fragment.
+   * @param {string} options.publicKeyMultibase - Multibase encoded Public Key.
+   * @param {string} [options.privateKeyMultibase] - Multibase Private Key.
```
Comment (mattcollier) [inline lib/X25519KeyAgreementKey2020.js]: why are "public", "private" and "key" being capped here?

File: lib/X25519KeyAgreementKey2020.js [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+function _multibaseDecode({text}) {
+  // drop the initial multibase 'z' prefix
+  return base58btc.decode(text.substr(1));
+}
+
+export class X25519KeyAgreementKey2020 extends LDKeyPair {
+  /**
+   * @param {object} options - Options hashmap.
+   * @param {string} options.controller - Controller DID or document url.
+   * @param {string} [options.id] - Key id, typically composed of controller
```
Comment (mattcollier) [inline lib/X25519KeyAgreementKey2020.js]:
```suggestion
   * @param {string} [options.id] - Key ID, typically composed of controller
```

File: lib/X25519KeyAgreementKey2020.js [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+   *   URL and key fingerprint as hash fragment.
+   * @param {string} options.publicKeyMultibase - Multibase encoded Public Key.
+   * @param {string} [options.privateKeyMultibase] - Multibase Private Key.
+   */
+  constructor(options = {}) {
+    super(options);
+    this.type = SUITE_ID;
+    this.publicKeyMultibase = options.publicKeyMultibase;
+    if(!this.publicKeyMultibase) {
+      throw TypeError('The "publicKeyMultibase" property is required.');
```
Comment (mattcollier) [inline lib/X25519KeyAgreementKey2020.js]:
```suggestion
      throw TypeError('The "publicKeyMultibase" parameter is required.');
```

File: lib/X25519KeyAgreementKey2020.js [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+    publicKeyBase58, privateKeyBase58, ...keyPairOptions
+  } = {}) {
+    let publicKeyMultibase, privateKeyMultibase;
+
+    if(publicKeyBase58) {
+      publicKeyMultibase = `z${publicKeyBase58}`;
+    }
+    if(privateKeyBase58) {
+      privateKeyMultibase = `z${privateKeyBase58}`;
+    }
```
Comment (mattcollier) [inline lib/X25519KeyAgreementKey2020.js]: A comment about the `z` prefix would be helpful to the reader unfamiliar with this multibase shortcut.

File: lib/X25519KeyAgreementKey2020.js [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+   *   key pair.
+   */
+  static fromEd25519VerificationKey2020({keyPair} = {}) {
+    if(!keyPair.publicKeyMultibase) {
+      throw new Error('Source public key is required to convert.');
+    }
+
+    if(!keyPair.publicKeyMultibase.startsWith('z')) {
+      throw new TypeError(
+        'Expecting source public Ed25519 2020 key to have base58btc encoding.'
```
Comment (mattcollier) [inline lib/X25519KeyAgreementKey2020.js]: I think `multibase` needs to be worked in here somewhere.  Same for private key on L150.

File: lib/X25519KeyAgreementKey2020.js [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+      exportedKey.publicKeyMultibase = this.publicKeyMultibase;
+    }
+    if(privateKey) {
+      exportedKey.privateKeyMultibase = this.privateKeyMultibase;
+    }
+    return exportedKey;
+  }
+
+  /**
+   * Generates and returns a multiformats encoded
```
Comment (mattcollier) [inline lib/X25519KeyAgreementKey2020.js]: I think describing this is `multibase` encoded would be better.

File: lib/X25519KeyAgreementKey2020.js [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+  static fromFingerprint({fingerprint} = {}) {
+    if(!fingerprint ||
+      !(typeof fingerprint === 'string' && fingerprint[0] === 'z')) {
+      throw new Error('`fingerprint` must be a multibase encoded string.');
+    }
+    const buffer = _multibaseDecode({text: fingerprint});
+
+    // buffer is: 0xec 0x01 <public key bytes>
+    if(buffer[0] !== 0xec || buffer[1] !== 0x01) {
+      throw new Error(`Unsupported Fingerprint Type: ${fingerprint}`);
```
Comment (mattcollier) [inline lib/X25519KeyAgreementKey2020.js]: The first part of this error makes me expect that after the colon will be a TYPE, bit's it's just echoing back the fingerprint that was passed in.  Something that helps the user correct the issue would be good.

File: lib/X25519KeyAgreementKey2020.js [hunk 10]
```diff
@@ -0,0 +1,375 @@
...
+    }
+
+    return new X25519KeyAgreementKey2020({
+      publicKeyMultibase: _multibaseEncode({bytes: buffer.slice(2)})
```
Comment (mattcollier) [inline lib/X25519KeyAgreementKey2020.js]: a comment about why the slice would be helpful.  Also, could that bit just be part of the helper?

File: lib/X25519KeyAgreementKey2020.js [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+    });
+  }
+
+  /**
+   * Derives a shared secret via a given public key, typically for use
+   * as one parameter for computing a shared key. It should not be used as
+   * a shared key itself, but rather input into a key derivation function (KDF)
```
Comment (mattcollier) [inline lib/X25519KeyAgreementKey2020.js]:
```suggestion
   * a shared key itself, but rather as an input into a key derivation function (KDF)
```

File: lib/X25519KeyAgreementKey2020.js [hunk 12]
```diff
@@ -0,0 +1,375 @@
...
+   * to produce a shared key.
+   *
+   * @param {object} [options={}] - Options hashmap.
+   * @param {LDKeyPair} options.publicKey - Remote key pair.
+   *
+   * @returns {string} Derived secret.
```
Comment (mattcollier) [inline lib/X25519KeyAgreementKey2020.js]: I believe we are still trying to be explicit wrt async stuffs
```suggestion
   * @returns {Promise<string>} Derived secret.
```

File: lib/X25519KeyAgreementKey2020.js [hunk 13]
```diff
@@ -0,0 +1,375 @@
...
+    let publicKeyBytes;
+    try {
+      publicKeyBytes = _multibaseDecode({text: this.publicKeyMultibase});
+    } catch(e) {
+      return {error: e, valid: false};
+    }
+    // validate the first buffer multicodec bytes 0xec 0x01
+    const valid = fingerprintBytes[0] === 0xec &&
+      fingerprintBytes[1] === 0x01 &&
+      publicKeyBytes.toString() === fingerprintBytes.slice(2).toString();
```
Comment (mattcollier) [inline lib/X25519KeyAgreementKey2020.js]: This is the second time I've seen this sort of code.  I think a well documented/commented helper would be good to DRY things up and provide explanation.

File: lib/crypto-nacl.js [hunk 14]
```diff
@@ -0,0 +1,14 @@
+/*!
+ * Copyright (c) 2021 Digital Bazaar, Inc. All rights reserved.
+ */
+import nacl from 'tweetnacl';
+
+export async function deriveSecret({privateKey, remotePublicKey}) {
+  return nacl.scalarMult(privateKey, remotePublicKey);
+}
+
+export async function generateKeyPair() {
```
Comment (mattcollier) [inline lib/crypto-nacl.js:15]: a comment that this is async for parity with the native implementation would be helpful.

**Review by mattcollier — APPROVED** [review]
No show stoppers.  Approving with lots of doc nits.

**Comment by mattcollier** [comment]
@dmitrizagidulin I initialized the repo on codecov.  Coverage is now failing because the coverage report is not being found.

File: lib/X25519KeyAgreementKey2020.js (hunk 3 above)
Comment (dmitrizagidulin) [reply lib/X25519KeyAgreementKey2020.js]: Note sure; fixed.

File: lib/X25519KeyAgreementKey2020.js (hunk 6 above)
Comment (dmitrizagidulin) [reply lib/X25519KeyAgreementKey2020.js]: Done, thanks.

File: lib/X25519KeyAgreementKey2020.js (hunk 7 above)
Comment (dmitrizagidulin) [reply lib/X25519KeyAgreementKey2020.js]: Rephrased.

File: lib/X25519KeyAgreementKey2020.js (hunk 8 above)
Comment (dmitrizagidulin) [reply lib/X25519KeyAgreementKey2020.js]: Rephrased.

File: lib/X25519KeyAgreementKey2020.js (hunk 9 above)
Comment (dmitrizagidulin) [reply lib/X25519KeyAgreementKey2020.js]: Rephrased, moved to helper.

File: lib/X25519KeyAgreementKey2020.js (hunk 10 above)
Comment (dmitrizagidulin) [reply lib/X25519KeyAgreementKey2020.js]: Added comment, moved to helper.

File: lib/X25519KeyAgreementKey2020.js (hunk 12 above)
Comment (dmitrizagidulin) [reply lib/X25519KeyAgreementKey2020.js]: Oversight; fixed.

File: lib/X25519KeyAgreementKey2020.js (hunk 13 above)
Comment (dmitrizagidulin) [reply lib/X25519KeyAgreementKey2020.js]: Np, moved to helper.

File: lib/crypto-nacl.js (hunk 14 above)
Comment (dmitrizagidulin) [reply lib/crypto-nacl.js:15]: Added.

**Pull request merged**